### ISSUE TITLE
Not panic if bind failed

### ISSUE THREAD

**Issue by zrll12** [actix/actix-web#2958]
This only exists on windows platforms.

## Expected Behavior
Panic if can not bind to port

## Current Behavior
continue runs, but can not really listen

## Steps to Reproduce (for bugs)
1. build
2. run at least 2

## Context
My code:
```rust
HttpServer::new(|| {
    App::new()
        .service(echo)
        .service(user::login_request)
        .service(user::register_request)
        .service(user::verify_email_request)
})
.bind(("0.0.0.0", unsafe { &CONFIG }["connection"]["serverPort"].as_i64().unwrap() as u16))
.expect("Can not bind server to port").run().await.expect("Can not start server");
```

## Your Environment
Windows 10 22H2 19045.2364

* Rust Version (I.e, output of `rustc -V`): rustc 1.68.0-nightly (77429957a 2023-01-01)
* Actix-* crate(s) Version: { version = "4.2.1", features = ["rustls"] }

**Comment by kaffarell** [comment]
Yep, could reproduce this on my (windows 11) machine.

Happens because of this (check the line 797, we allow a re-use of the address (port)):
https://github.com/actix/actix-web/blob/17218dc6c88848938cebc560deafbf1c2184fa56/actix-web/src/server.rs#L793-L803

I don't know if this is intended, but it surely confuses a lot of people because two instances bind to the same address, but only one receives the requests.

imo we should set this to false.

@robjtede ?

**Comment by imor** [comment]
The `SO_REUSEADDR` option impacts two aspects of a socket. The first is whether two sockets with conflicting addresses are allowed to bind or not. The second is if a socket in `TIME_WAIT` state prevents a new socket from binding to the same address. This is discussed in detail in [this SO answer](https://stackoverflow.com/questions/14388706/how-do-so-reuseaddr-and-so-reuseport-differ/14388707#14388707). So just setting this to false will alter the behaviour for both of the above aspects. I tested on both Windows (Win10) and Linux (Ubuntu 22.04) by inducing the server socket into the `TIME_WAIT` state and trying to bind. With `SO_REUSEADDR` set to true, both OSs allowed binding but with false only Windows allowed to bind.

So I think a better option is to set this to true only on non-Windows platforms:

```rust
fn create_tcp_listener(addr: net::SocketAddr, backlog: u32) -> io::Result<net::TcpListener> {
    use socket2::{Domain, Protocol, Socket, Type};
    let domain = Domain::for_address(addr);
    let socket = Socket::new(domain, Type::STREAM, Some(Protocol::TCP))?;
    #[cfg(not(windows))]
    {
        socket.set_reuse_address(true)?;
    }
    socket.bind(&addr.into())?;
    // clamp backlog to max u32 that fits in i32 range
    let backlog = cmp::min(backlog, i32::MAX as u32) as i32;
    socket.listen(backlog)?;
    Ok(net::TcpListener::from(socket))
}
```
This will make sure that Windows behaviour better matches with other platforms like Linux if exact same ip:port is used by two processes. Note that even after my suggested change there is a difference in behaviour between Windows and Linux if one process binds to a wildcard (e.g. 0.0.0.0:8080) and another to a specific address (127.0.0.1:8080). In this case the bind succeeds on Windows but fails on Linux, but this is more of an edge case so should be ok.

The above suggested change should also fix #1335 and #1913 

One thing to keep in mind is that this will be a breaking change on Windows. So not sure how to introduce it. Maybe in next major version? @robjtede any ideas?

**Comment by bjones1** [comment]
@imor, thanks for this fantastic analysis! I concur -- I primarily use actix on Windows; I've been bitten by the current behavior several times. In my case, I left my server running, then also ran unit tests, which start another copy of the server. My tests failed in difficult-to-debug ways -- because **both** servers bound to the same IP address and port, but the first server (production, not unit test) handled all the traffic generated by the tests. Preventing two servers from binding to the same address is a great fix!

I tested on Windows by commenting out the line `socket.set_reuse_address(true)?;` and this prevented multiple bindings to the same port. I therefore agree with this fix and would like to see it merged.

I'd also like to thank @robjtede first for creating and maintain actix -- it's been a pleasure to work with this software! In addition, his responses on Discord to my initial questions on this topic led me here.

I thought I'd pass on a few observations I made when initially researching this topic, which are also posted on Discord, hoping they'll be helpful to others:

- On Windows, use [SO_EXCLUSIVEADDRUSE](https://learn.microsoft.com/en-us/windows/win32/winsock/so-exclusiveaddruse) instead of the SO_REUSEADDR due to Windows bugs; the SO article linked above by @imor covers this nicely. See also more [Windows docs](https://learn.microsoft.com/en-us/windows/win32/winsock/using-so-reuseaddr-and-so-exclusiveaddruse). I'm not sure which flavor the socket2 library uses. (Of course, the proposed fix does not use this flag, making this somewhat irrelevant.)
